includes/payment-methods/class-upe-payment-gateway.php

```diff
@@ -0,0 +1,301 @@
+							wp_create_nonce( 'wcpay_update_order_status_nonce' )
+						);
+						wp_safe_redirect( $redirect_url );
+						exit;
+					}
+				}
+			}
+		} catch ( Exception $e ) {
+			Logger::log( 'Error: ' . $e->getMessage() );
+
+			/* translators: localized exception message */
+			$order->update_status( 'failed', sprintf( __( 'UPE payment failed: %s', 'woocommerce-payments' ), $e->getMessage() ) );
+
+			wc_add_notice( $e->getMessage(), 'error' );
+			wp_safe_redirect( wc_get_checkout_url() );
+			exit;
+		}
+	}
+
+	/**
+	 * Renders the UPE input fields needed to get the user's payment information on the checkout page.
+	 *
+	 * We also add the JavaScript which drives the UI.
+	 */
+	public function payment_fields() {
+		try {
+			$display_tokenization = $this->supports( 'tokenization' ) && is_checkout();
+
+			$payment_fields                  = $this->get_payment_fields_js_config();
+			$payment_fields['paymentIntent'] = $this->create_payment_intent();
+			wp_localize_script( 'WCPAY_CHECKOUT', 'wcpay_config', $payment_fields );
+			wp_enqueue_script( 'WCPAY_CHECKOUT' );
+
+			$prepared_customer_data = $this->get_prepared_customer_data();
+			if ( ! empty( $prepared_customer_data ) ) {
+				wp_localize_script( 'WCPAY_CHECKOUT', 'wcpayCustomerData', $prepared_customer_data );
+			}
+
+			wp_enqueue_style(
+				'WCPAY_CHECKOUT',
+				plugins_url( 'dist/checkout.css', WCPAY_PLUGIN_FILE ),
+				[],
+				WC_Payments::get_file_version( 'dist/checkout.css' )
+			);
+
+			// Output the form HTML.
+			?>
+			<?php if ( ! empty( $this->get_description() ) ) : ?>
+				<p><?php echo wp_kses_post( $this->get_description() ); ?></p>
+			<?php endif; ?>
+
+			<?php if ( $this->is_in_test_mode() ) : ?>
+				<p class="testmode-info">
+				<?php
+					echo WC_Payments_Utils::esc_interpolated_html(
+						/* translators: link to Stripe testing page */
+						__( '<strong>Test mode:</strong> you will be redirected to a Stripe test page to authorize payment.', 'woocommerce-payments' ),
+						[
+							'strong' => '<strong>',
+						]
+					);
+				?>
+				</p>
+			<?php endif; ?>
+
+			<?php
+			if ( $display_tokenization ) {
+				$this->tokenization_script();
+				echo $this->saved_payment_methods(); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+			}
+			?>
+
+			<fieldset id="wc-<?php echo esc_attr( $this->id ); ?>-upe-form" class="wc-upe-form wc-payment-form">
+				<div id="wcpay-upe-element"></div>
+				<div id="wcpay-upe-errors" role="alert"></div>
+				<input id="wcpay-payment-method-upe" type="hidden" name="wcpay-payment-method-upe" />
+
+			<?php
+			if ( $this->is_saved_cards_enabled() ) {
+				$force_save_payment = ( $display_tokenization && ! apply_filters( 'wc_payments_display_save_payment_method_checkbox', $display_tokenization ) ) || is_add_payment_method_page();
+				$this->save_payment_method_checkbox( $force_save_payment );
+			}
+			?>
+
+			</fieldset>
+			<?php
+		} catch ( Exception $e ) {
+			// Output the error message.
```

Comment:
```suggestion
			Logger::log( 'Error: ' . $e->getMessage() );
```
Probably we should at least log an errors.

Reply:
Agreed. Sorted.